src/Features/CSharp/Portable/Diagnostics/Analyzers/TypeSyntaxSimplifierWalker_Scopes.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.SimplifyTypeNames
+{
+    internal partial class TypeSyntaxSimplifierWalker : CSharpSyntaxWalker
+    {
+        private readonly CSharpSimplifyTypeNamesDiagnosticAnalyzer _analyzer;
+        private readonly SemanticModel _semanticModel;
+        private readonly OptionSet _optionSet;
+        private readonly bool _preferPredefinedTypeInDecl;
+        private readonly bool _preferPredefinedTypeInMemberAccess;
+        private readonly CancellationToken _cancellationToken;
+
+        /// <summary>
+        /// Set of all the type names that this compilation knows about (both through source and
+        /// metadata).  This is used so that when we see <c>Y.Z</c> we can know if <c>Y</c> could be
+        /// a type and thus if we should try to simplify that to <c>X.Z</c> (if <c>X</c> is a base
+        /// class of <c>Y</c>).
+        /// </summary>
+        private readonly HashSet<string> _compilationTypeNames;
+
+        /// <summary>
+        /// Set of type and namespace names that have an alias associated with them.  i.e. if the
+        /// user has <c>using X = System.DateTime</c>, then <c>DateTime</c> will be in this set.
+        /// This is used so we can easily tell if we should try to simplify some identifier to an
+        /// alias when we encounter it.
+        /// </summary>
+        private HashSet<string> _aliasedSymbolNames;
```

Comment:
❔ Do you know how much we gain from this specifically? If the benefits are substantial, do you know why #23556 didn't already fix it?